packages/core/content-manager/admin/src/pages/EditView/components/FormInputs/BlocksInput/BlocksToolbar.tsx

```diff
@@ -538,32 +667,104 @@ const BlocksToolbar = () => {
 
   const isButtonDisabled = checkButtonDisabled();
 
+  /**
+   * Observed items are ones that may or may not be visible in the toolbar, depending on the
+   * available space. They provide two render props:
+   * - renderInToolbar: for when we try to render the item in the toolbar (may be hidden)
+   * - renderInMenu: for when the item didn't fit in the toolbar and is relegated to the "more" menu
+   */
+  const observedItems = [
+    ...Object.entries(modifiers).map(([name, modifier]) => ({
+      renderInToolbar: () => (
+        <ToolbarButton
+          key={name}
+          name={name}
+          icon={modifier.icon}
+          label={modifier.label}
+          isActive={modifier.checkIsActive(editor)}
+          handleClick={() => modifier.handleToggle(editor)}
+          disabled={isButtonDisabled}
+        />
+      ),
+      renderInMenu: () => (
+        <Menu.Item onClick={() => modifier.handleToggle(editor)}>
+          {formatMessage(modifier.label)}
+        </Menu.Item>
+      ),
+    })),
+    {
+      renderInToolbar: () => <LinkButton disabled={isButtonDisabled} location="toolbar" />,
+      renderInMenu: () => <LinkButton disabled={isButtonDisabled} location="menu" />,
+    },
+    {
+      // List buttons can only be rendered together when in the toolbar
+      renderInToolbar: () => (
+        <Flex direction="row" gap={1}>
+          <Separator />
+          <Toolbar.ToggleGroup type="single" asChild>
+            <Flex gap={1}>
+              <ListButton block={blocks['list-unordered']} format="unordered" location="toolbar" />
+              <ListButton block={blocks['list-ordered']} format="ordered" location="toolbar" />
+            </Flex>
+          </Toolbar.ToggleGroup>
+        </Flex>
+      ),
+      renderInMenu: () => [
+        <ListButton
+          block={blocks['list-unordered']}
+          format="unordered"
+          location="menu"
+          key="unordered"
+        />,
+        <ListButton
+          block={blocks['list-ordered']}
+          format="ordered"
+          location="menu"
+          key="ordered"
+        />,
+      ],
+    },
+  ] as Array<ObservedItem>;
```

Comment:
This makes the component quite long, maybe we could extract it into a useFlexibleItems hook so it's easier to read

Reply:
Sure sounds good, not sure about the name though 🤔. I think I prefer the notion of `observed` since they are designed to be used with the intersection observer.

Reply:
yes observed is better. But actually if we end up abstracting the logic further I'm not sure it will be needed since the component code will be a lot shorter